.cbmc-batch/jobs/aws_priority_queue_push/aws_priority_queue_push_harness.c

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/priority_queue.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void aws_priority_queue_push_harness() {
+    /* Data structure */
+    struct aws_priority_queue queue;
+
+    /* Assumptions */
+    __CPROVER_assume(aws_priority_queue_is_bounded(&queue, MAX_INITIAL_ITEM_ALLOCATION, MAX_ITEM_SIZE));
+    ensure_priority_queue_has_allocated_members(&queue);
+    __CPROVER_assume(aws_priority_queue_is_valid(&queue));
+    void *item = can_fail_malloc(queue.container.item_size);
+
+    /* Assume the function preconditions */
+    __CPROVER_assume(item && AWS_MEM_IS_READABLE(item, queue.container.item_size));
```

Comment:
I'm not asking for changes here. But out of curiosity: is it documented somewhere that (for example) `item` should be non-NULL as a precondition?

I had an idea (for the distant future) that we should generate a human-readable report from these harnesses. We'd scrape all the __CPROVER_assumes that happen before the function call and generate a cute report that says "we've proved foo, assuming bar and baz". But I suppose this is slightly less useful for c-common as opposed to FreeRTOS, if it's already documented somewhere.

Reply:
At the moment, `AWS_PRECONDITION` and `AWS_POSTCONDITION` in the function code serve as documentation (especially after #369, which allows for explanations in pre/postconditions, was merged). However, as you noted the "real" assumptions are written in the harnesses and thus they could "drift" out-of-sync from preconditions. It would be really nice (as future work) to generate documentation from `__CPROVER_assume`s or to find a way to keep assumes and preconditions in sync.